critwall/src/main/java/org/opencadc/critwall/FileSyncJob.java

```diff
@@ -354,37 +352,31 @@ private void doSync() {
 
         TransferReader reader = new TransferReader();
         Transfer t = reader.read(post.getInputStream(), null);
-        List<String> urlStrList = t.getAllEndpoints();
-        log.debug("endpoints returned: " + urlStrList);
-
-        List<URL> urlList = new ArrayList<>();
-
-        // Create URL list to return
-        for (final String s : urlStrList) {
-            try {
-                urlList.add(new URL(s));
-            } catch (MalformedURLException mue) {
-                log.info("malformed URL returned from transfer negotiation: " + s + " skipping... ");
-            }
-        }
-
-        return urlList;
+        return t.getProtocols();
     }
 
-    private StorageMetadata syncArtifact(Artifact a, List<URL> urls)
-        throws ByteLimitExceededException, StorageEngageException, InterruptedException, WriteException, IllegalArgumentException {
+    private StorageMetadata syncArtifact(Artifact a, List<Protocol> urls)
+        throws ByteLimitExceededException, StorageEngageException, InterruptedException, 
+        WriteException, IllegalArgumentException, MalformedURLException, PrivilegedActionException {
 
         StorageMetadata storageMeta = null;
-        Iterator<URL> urlIterator = urls.iterator();
+        Iterator<Protocol> urlIterator = urls.iterator();
 
         while (urlIterator.hasNext()) {
-            URL u = urlIterator.next();
+            Protocol p = urlIterator.next();
+            URL u = new URL(p.getEndpoint());
             log.debug("trying " + u);
 
             boolean postPrepare = false;
             try {
                 ByteArrayOutputStream dest = new ByteArrayOutputStream();
                 HttpGet get = new HttpGet(u, dest);
+                if (p.getSecurityMethod() == null || p.getSecurityMethod().equals(Standards.getSecurityMethod(AuthMethod.ANON))) {
+                    get = Subject.doAs(anonSubject, new NestedAction(u));
+                } else {
+                    get.prepare();
+                }
+                
```

Comment:
Ack!!! Those timeouts *must* be set before you make the call - right after creating the HttpGet instance! They can't do any good after the call.

And I explicitly said to use the HttpGet instance created here, not create a new one in the action and return it: for the very reason that it makes this kind of mess where important state is not called or not known. Why create an instance (line 373) then immediately ignore it, pass the raw url into the action, and create a new one? 

And you still have another call to get.prepare() on line 382.

That's 4 bugs caused by one cut&paste!

Reply:
Sorry about the careless errors. Fixed, I think.